File: solver.py
"""
Newton-Raphson solver to find fixed rate
"""
from scipy import optimize

def npv(couponRate, notional, coupons, targetNpv):
    legNpv = sum([notional * couponRate * cf.NbYears * cf.EndDiscountFactor for cf in coupons[0:-1]])
    legNpv += (notional * ((1 + couponRate) * coupons[-1].NbYears) * coupons[-1].EndDiscountFactor)
    return targetNpv - legNpv
    
def solve(notional, coupons, targetNpv):
    x0 = 0.05
    return optimize.newton(npv, x0, args=(notional, coupons, targetNpv))

File: csharp/vanilla_swap/vanilla_swap/Solver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using mn = MathNet.Numerics;

namespace vanilla_swap
{
    public static class Solver
    {
        public static double Solve(double notional, IList<Coupon> coupons, double targetNpv)
        {
            return mn.RootFinding.RobustNewtonRaphson.FindRoot(
                r => 
                    targetNpv - (coupons.Sum(x => notional * r * x.NbYears * x.EndDiscountFactor) + (notional * coupons.Last().NbYears * coupons.Last().EndDiscountFactor)),
                r =>
                    -(coupons.Sum(x => notional * x.NbYears * x.EndDiscountFactor)),
                -1, 1);
        }
    }
}
